client/src/controllers/CleanController.ts

```diff
@@ -0,0 +1,180 @@
+ * <input type="text" name="url-path" data-controller="w-slug" data-action="change->w-slug#urlify" />
+ * <input type="text" name="url-path-with-unicode" data-controller="w-slug" data-w-slug-allow-unicode="true" data-action="change->w-slug#urlify" />
+ * ```
+ */
+export class CleanController extends Controller<HTMLInputElement> {
+  static values = {
+    allowUnicode: { default: false, type: Boolean },
+    trim: { default: false, type: Boolean },
+  };
+
+  /**
+   * If true, unicode values in the cleaned values will be allowed.
+   * Otherwise unicode values will try to be transliterated.
+   * @see `WAGTAIL_ALLOW_UNICODE_SLUGS` in settings
+   */
+  declare readonly allowUnicodeValue: boolean;
+  /** If true, value will be trimmed in all clean methods before being processed by that method. */
+  declare readonly trimValue: boolean;
+
+  /**
+   * Writes the new value to the element & dispatches the applied event.
+   *
+   * @fires CleanController#applied - If a change applied to the input value, this event is dispatched.
+   *
+   * @event CleanController#applied
+   * @type {CustomEvent}
+   * @property {string} name - `w-slug:applied` | `w-clean:applied`
+   * @property {Object} detail
+   * @property {string} detail.action - The action that was applied (e.g. 'urlify' or 'slugify').
+   * @property {string} detail.cleanValue - The the cleaned value that is applied.
+   * @property {string} detail.sourceValue - The original value.
+   */
+  applyUpdate(action: Actions, cleanValue: string, sourceValue?: string) {
+    this.element.value = cleanValue;
+    this.dispatch('applied', {
+      cancelable: false,
+      detail: { action, cleanValue, sourceValue },
+    });
+  }
+
+  /**
+   * Allow for a comparison value to be provided so that a dispatched event can be
+   * prevented. This provides a way for other events to interact with this controller
+   * to block further updates if a value is not in sync.
+   * By default it will compare to the slugify method, this can be overridden by providing
+   * either a Stimulus param value on the element or the event's detail.
+   */
+  compare(
+    event: CustomEvent<{ compareAs?: Actions; value: string }> & {
+      params?: { compareAs?: Actions };
+    },
+  ) {
+    // do not attempt to compare if the field is empty
+    if (!this.element.value) return true;
+
+    const compareAs =
+      event.detail?.compareAs || event.params?.compareAs || Actions.Slugify;
+
+    const compareValue = this[compareAs](
+      { detail: { value: event.detail?.value || '' } },
+      { ignoreUpdate: true },
+    );
+
+    const valuesAreSame = this.compareValues(compareValue, this.element.value);
+
+    if (!valuesAreSame) {
+      event?.preventDefault();
+    }
+
+    return valuesAreSame;
+  }
+
+  /**
+   * Compares the provided strings, ensuring the values are the same.
+   */
+  compareValues(...values: string[]): boolean {
+    return new Set(values.map((value: string) => `${value}`)).size === 1;
+  }
```

Comment:
I don’t understand when we would sync more than two fields, and when those fields’ values wouldn’t be strings 🤔